File: app/src/main/java/com/imangazalievm/bubbble/presentation/mvp/userprofile/UserDetailsPresenter.java
package com.imangazalievm.bubbble.presentation.mvp.userprofile;

import com.arellomobile.mvp.InjectViewState;
import com.arellomobile.mvp.MvpPresenter;
import com.imangazalievm.bubbble.domain.global.exceptions.NoNetworkException;
import com.imangazalievm.bubbble.domain.global.models.User;
import com.imangazalievm.bubbble.domain.userprofile.UserDetailsInteractor;
import com.imangazalievm.bubbble.presentation.mvp.global.SchedulersProvider;
import com.imangazalievm.bubbble.presentation.utils.DebugUtils;

import javax.inject.Inject;

@InjectViewState
public class UserDetailsPresenter extends MvpPresenter<UserDetailsView> {

    private UserDetailsInteractor userDetailsInteractor;
    private SchedulersProvider schedulersProvider;
    private long userId;
    private User user;

    @Inject
    public UserDetailsPresenter(
            UserDetailsInteractor userDetailsInteractor,
            SchedulersProvider schedulersProvider,
            long userId
    ) {
        this.userDetailsInteractor = userDetailsInteractor;
        this.schedulersProvider = schedulersProvider;
        this.userId = userId;


    }

    @Override
    protected void onFirstViewAttach() {
        super.onFirstViewAttach();

        loadUser();
    }

    private void loadUser() {
        getViewState().showLoadingProgress();
        userDetailsInteractor.getUser(userId)
                .observeOn(schedulersProvider.ui())
                .subscribe(this::onUserLoadSuccess, this::onUserLoadError);
    }

    private void onUserLoadSuccess(User user) {
        this.user = user;
        getViewState().hideLoadingProgress();
        getViewState().showUserInfo(user);
    }

    private void onUserLoadError(Throwable throwable) {
        if (throwable instanceof NoNetworkException) {
            getViewState().hideLoadingProgress();
            getViewState().showNoNetworkLayout();
        } else {
            DebugUtils.showDebugErrorMessage(throwable);
        }
    }

    public void retryLoading() {
        getViewState().hideNoNetworkLayout();
        loadUser();
    }

    public void onUserTwitterButtonClicked() {
        getViewState().openInBrowser(user.getLinks().getTwitter());
    }

    public void onUserWebsiteButtonClicked() {
        getViewState().openInBrowser(user.getLinks().getWeb());
    }

}

File: app/src/main/java/com/imangazalievm/bubbble/presentation/userprofile/details/UserDetailsPresenter.kt
package com.imangazalievm.bubbble.presentation.userprofile.details

import com.arellomobile.mvp.InjectViewState
import com.imangazalievm.bubbble.data.global.network.exceptions.NoNetworkException
import com.imangazalievm.bubbble.domain.global.models.User
import com.imangazalievm.bubbble.domain.userprofile.UserDetailsInteractor
import com.imangazalievm.bubbble.presentation.global.mvp.BasePresenter
import javax.inject.Inject

@InjectViewState
class UserDetailsPresenter @Inject constructor(
    private val userDetailsInteractor: UserDetailsInteractor,
    private val userId: Long
) : BasePresenter<UserDetailsView>() {

    private lateinit var user: User
    private val isUserLoaded: Boolean
        get() = ::user.isInitialized

    override fun onFirstViewAttach() {
        super.onFirstViewAttach()
        loadUser()
    }

    private fun loadUser() = launchSafe {
        viewState.showLoadingProgress()
        try {
            this@UserDetailsPresenter.user = userDetailsInteractor.getUser(userId)
            viewState.showUserInfo(user)
        } catch (e: NoNetworkException) {
            viewState.showNoNetworkLayout()
        } finally {
            viewState.hideLoadingProgress()
        }
    }

    fun retryLoading() {
        viewState.hideNoNetworkLayout()
        loadUser()
    }

    fun onUserTwitterButtonClicked() {
        if (!isUserLoaded) return
        viewState.openInBrowser(user.links.twitter)
    }

    fun onUserWebsiteButtonClicked() {
        if (!isUserLoaded) return
        viewState.openInBrowser(user.links.web)
    }
}